modules/gapi/src/executor/gstreamingexecutor.cpp

```diff
@@ -0,0 +1,676 @@
+                    isl_outputs[id] = { r, cv::GRunArgP(rr) };
+                }
+                break;
+            default:
+                cv::util::throw_error(std::logic_error("Unsupported GShape"));
+                break;
+            }
+        }
+        // Now ask Island to execute on this data
+        island->run(std::move(isl_inputs), std::move(isl_outputs));
+
+        // Once executed, dispatch our results down to the pipeline.
+        for (auto &&it : ade::util::zip(ade::util::toRange(out_queues),
+                                        ade::util::toRange(out_data)))
+        {
+            for (auto &&q : std::get<0>(it))
+            {
+                // FIXME: FATAL VARIANT ISSUE!!
+                const auto tmp = std::get<1>(it);
+                q->push(Cmd{tmp});
+            }
+        }
+    }
+}
+
+void collectorThread(std::vector<Q*> in_queues,
+                     Q&              out_queue)
+{
+    while (true)
+    {
+        cv::GRunArgs this_result(in_queues.size());
+        std::size_t stops = 0u;
+        for (auto &&it : ade::util::indexed(in_queues))
+        {
+            Cmd cmd;
+            ade::util::value(it)->pop(cmd);
+            if (cmd.index() == cmd.index_of<Stop>()) {
+                stops++;
+            } else {
+                // FIXME: MOVE_PROBLEM
+                const cv::GRunArg &in_arg = cv::util::get<cv::GRunArg>(cmd);
+                this_result[ade::util::index(it)] = in_arg;
+                // FIXME: Check for other message types.
+            }
+        }
+        if (stops > 0)
+        {
+            GAPI_Assert(stops == in_queues.size());
+            out_queue.push(Cmd{Stop{}});
+            return;
+        }
+        out_queue.push(Cmd{this_result});
+    }
+}
+} // anonymous namespace
+
+cv::gimpl::GStreamingExecutor::GStreamingExecutor(std::unique_ptr<ade::Graph> &&g_model)
+    : m_orig_graph(std::move(g_model))
+    , m_island_graph(GModel::Graph(*m_orig_graph).metadata()
+                     .get<IslandModel>().model)
+    , m_gm(*m_orig_graph)
+    , m_gim(*m_island_graph)
+{
+    // NB: Right now GIslandModel is acyclic, and all the below code assumes that.
+    // NB: This naive execution code is taken from GExecutor nearly "as-is"
+
+    const auto proto = m_gm.metadata().get<Protocol>();
+    m_emitters      .resize(proto.in_nhs.size());
+    m_emitter_queues.resize(proto.in_nhs.size());
+    m_sinks         .resize(proto.out_nhs.size());
+    m_sink_queues   .resize(proto.out_nhs.size());
+
+    // Very rough estimation to limit internal queue sizes.
+    // Pipeline depth is equal to number of its (pipeline) steps.
+    const auto queue_capacity = std::count_if
+        (m_gim.nodes().begin(),
+         m_gim.nodes().end(),
+         [&](ade::NodeHandle nh) {
+            return m_gim.metadata(nh).get<NodeKind>().k == NodeKind::ISLAND;
+         });
```

Comment:
Is this counting routine will also work for pipelines with parallel islands?
I mean if islands number will be what we need?

Reply:
Even if islands are topologically parallel, it doesn't mean those run on different devices.

Just counting the total number of islands is a nice initial estimation heuristic. In the worst case, neither of our queues will be full - nothing criminal here.